opnode/test/system_test.go

```diff
@@ -416,45 +244,159 @@ loop:
+	require.Nil(t, err, "Waiting for L2 tx on verifier")
 
-		receipt, err := l2Client.TransactionReceipt(context.Background(), tx.Hash())
-		if receipt != nil && err == nil {
-			l2IncludedBlock = receipt.BlockNumber
-			break lastLoop
-		} else if err != nil && !errors.Is(err, ethereum.NotFound) {
-			require.Nil(t, err)
-		}
-	}
-
-	verifBlock, err := l2Client.BlockByNumber(context.Background(), l2IncludedBlock)
+	// Verify blocks match after batch submission on verifiers and sequencers
+	verifBlock, err := l2Verif.BlockByNumber(context.Background(), receipt.BlockNumber)
 	require.Nil(t, err)
-	seqBlock, err := l2SequencerClient.BlockByNumber(context.Background(), l2IncludedBlock)
+	seqBlock, err := l2Seq.BlockByNumber(context.Background(), receipt.BlockNumber)
 	require.Nil(t, err)
 	require.Equal(t, verifBlock.Hash(), seqBlock.Hash(), "Verifier and sequencer blocks not the same after including a batch tx")
 
 }
+
+func TestMissingBatchE2E(t *testing.T) {
+	log.Root().SetHandler(log.DiscardHandler()) // Comment this out to see geth l1/l2 logs
+
+	cfg := defaultSystemConfig(t)
+	// Specifically set batch submitter balance to stop batches from being included
+	cfg.Premine[bssHDPath] = 0
+	// Don't pollute log with expected "Error submitting batch" logs
+	cfg.Loggers["sequencer"] = testlog.Logger(t, log.LvlCrit)
+
+	sys, err := cfg.start()
+	require.Nil(t, err, "Error starting up system")
+	defer sys.Close()
+
+	l2Seq := sys.Clients["sequencer"]
+	l2Verif := sys.Clients["verifier"]
+
+	// Transactor Account
+	ethPrivKey, err := sys.wallet.PrivateKey(accounts.Account{
+		URL: accounts.URL{
+			Path: transactorHDPath,
+		},
+	})
+	require.Nil(t, err)
+
+	// Submit TX to L2 sequencer node
+	toAddr := common.Address{0xff, 0xff}
+	tx := types.MustSignNewTx(ethPrivKey, types.LatestSignerForChainID(cfg.L2ChainID), &types.DynamicFeeTx{
+		ChainID:   cfg.L2ChainID,
+		Nonce:     0,
+		To:        &toAddr,
+		Value:     big.NewInt(1_000_000_000),
+		GasTipCap: big.NewInt(10),
+		GasFeeCap: big.NewInt(200),
+		Gas:       21000,
+	})
+	err = l2Seq.SendTransaction(context.Background(), tx)
+	require.Nil(t, err, "Sending L2 tx to sequencer")
+
+	// Let it show up on the unsafe chain
+	receipt, err := waitForTransaction(tx.Hash(), l2Seq, 6*time.Second)
+	require.Nil(t, err, "Waiting for L2 tx on sequencer")
+
+	// Wait until the block it was first included in shows up in the safe chain on the verifier
+	_, err = waitForBlock(receipt.BlockNumber, l2Verif, 4*time.Second)
+	require.Nil(t, err, "Waiting for block on verifier")
+
+	// Assert that the transaction is not found on the verifier
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	_, err = l2Verif.TransactionReceipt(ctx, tx.Hash())
+	require.Equal(t, ethereum.NotFound, err, "Found transaction in verifier when it should not have been included")
+
+	// Assert that the reconciliation process did an L2 reorg on the sequencer to remove the invalid block
```

Comment:
This reorgs out because the L1 clique node automines blocks?

Reply:
It's because sequencer included transactions in the L2 chain but failed to submit them to L1. The unsafe portion of the L2 chain needs to be reorged here.